File: backend/simulation/themes/epidemic.py
"""
Epidemic Theme
States: healthy → exposed → infected → recovered → dead
"""
import random

STATES = ["healthy", "exposed", "infected", "recovered", "dead"]
INITIAL_STATE = "healthy"
SEED_STATE = "infected"
SEED_FRACTION = 0.03

TRANSMISSION_RATE = 0.4
INCUBATION_TICKS = 3
RECOVERY_RATE = 0.08
DEATH_RATE = 0.01


def seed_states(agents: dict) -> None:
    ids = list(agents.keys())
    random.shuffle(ids)
    seed_count = max(1, int(len(ids) * SEED_FRACTION))
    for i, aid in enumerate(ids):
        agents[aid].state = SEED_STATE if i < seed_count else INITIAL_STATE


def rule_decision(agent, neighbor_states: list[str]) -> tuple[str, str]:
    state = agent.state
    traits = agent.traits()

    if state == "healthy":
        infected_neighbors = neighbor_states.count("infected") + neighbor_states.count("exposed")
        total = len(neighbor_states) or 1
        pressure = infected_neighbors / total
        immune_factor = 1.0 - traits["stubbornness"] / 200.0
        if random.random() < pressure * TRANSMISSION_RATE * immune_factor:
            return "exposed", f"contacted infected neighbor (pressure {pressure:.0%})"
        return "healthy", "avoided infection"

    elif state == "exposed":
        credulity_factor = traits["credulity"] / 100.0
        if random.random() < (0.2 + credulity_factor * 0.1):
            return "infected", "incubation period ended — symptomatic"
        return "exposed", "still incubating"

    elif state == "infected":
        if random.random() < DEATH_RATE:
            return "dead", "succumbed to illness"
        if random.random() < RECOVERY_RATE:
            return "recovered", "immune system prevailed"
        return "infected", "still sick"

    elif state == "recovered":
        return "recovered", "immune — recovered"

    elif state == "dead":
        return "dead", "deceased"

    return state, "no change"


def is_ambiguous(agent, neighbor_states: list[str]) -> bool:
    if not neighbor_states or agent.state not in ("healthy", "exposed"):
        return False
    infected = neighbor_states.count("infected") + neighbor_states.count("exposed")
    pressure = infected / len(neighbor_states)
    return 0.25 <= pressure <= 0.65


VALID_STATES = STATES
THEME_NAME = "Epidemic"
THEME_DESCRIPTION = "Agents get sick, recover, or die as disease spreads through contact networks."
STATE_COLORS = {
    "healthy": "#22c55e",
    "exposed": "#f59e0b",
    "infected": "#ef4444",
    "recovered": "#3b82f6",
    "dead": "#374151",
}

File: backend-node/src/simulation/themes/epidemic.ts
import type { Agent } from '../agent.js';
import type { Decision } from '../../types.js';
import { getTraits } from '../agent.js';

export const THEME_NAME = 'Epidemic';
export const THEME_DESCRIPTION = 'Agents get sick, recover, or die as disease spreads through contact networks.';
export const VALID_STATES = ['healthy', 'exposed', 'infected', 'recovered', 'dead'];
export const INITIAL_STATE = 'healthy';
export const SEED_STATE = 'infected';
export const SEED_FRACTION = 0.03;
export const EMOJI = '🦠';
export const DIFFICULTY: 'simple' | 'medium' | 'complex' = 'simple';

export const STATE_COLORS: Record<string, string> = {
  healthy: '#22c55e',
  exposed: '#f59e0b',
  infected: '#ef4444',
  recovered: '#3b82f6',
  dead: '#374151',
};

const TRANSMISSION_RATE = 0.4;
const RECOVERY_RATE = 0.08;
const DEATH_RATE = 0.01;

export function seedStates(agents: Map<string, Agent>): void {
  const ids = Array.from(agents.keys());
  shuffleArray(ids);
  const seedCount = Math.max(1, Math.floor(ids.length * SEED_FRACTION));
  
  ids.forEach((id, i) => {
    const agent = agents.get(id)!;
    agent.state = i < seedCount ? SEED_STATE : INITIAL_STATE;
  });
}

export function ruleDecision(agent: Agent, neighborStates: string[]): Decision {
  const state = agent.state;
  const traits = getTraits(agent);

  if (state === 'healthy') {
    const infectedNeighbors = neighborStates.filter(
      s => s === 'infected' || s === 'exposed'
    ).length;
    const total = neighborStates.length || 1;
    const pressure = infectedNeighbors / total;
    const immuneFactor = 1.0 - traits.stubbornness / 200.0;
    
    if (Math.random() < pressure * TRANSMISSION_RATE * immuneFactor) {
      return ['exposed', `contacted infected neighbor (pressure ${Math.round(pressure * 100)}%)`];
    }
    return ['healthy', 'avoided infection'];
  }

  if (state === 'exposed') {
    const credulityFactor = traits.credulity / 100.0;
    if (Math.random() < 0.2 + credulityFactor * 0.1) {
      return ['infected', 'incubation period ended — symptomatic'];
    }
    return ['exposed', 'still incubating'];
  }

  if (state === 'infected') {
    if (Math.random() < DEATH_RATE) {
      return ['dead', 'succumbed to illness'];
    }
    if (Math.random() < RECOVERY_RATE) {
      return ['recovered', 'immune system prevailed'];
    }
    return ['infected', 'still sick'];
  }

  if (state === 'recovered') {
    return ['recovered', 'immune — recovered'];
  }

  if (state === 'dead') {
    return ['dead', 'deceased'];
  }

  return [state, 'no change'];
}

export function isAmbiguous(agent: Agent, neighborStates: string[]): boolean {
  if (!neighborStates.length || !['healthy', 'exposed'].includes(agent.state)) {
    return false;
  }
  const infected = neighborStates.filter(
    s => s === 'infected' || s === 'exposed'
  ).length;
  const pressure = infected / neighborStates.length;
  return pressure >= 0.25 && pressure <= 0.65;
}

function shuffleArray<T>(array: T[]): void {
  for (let i = array.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [array[i], array[j]] = [array[j], array[i]];
  }
}
